make changes to allow optimization for both CPU and Memory configurations for GCP Cloud Functions V2

hdu3 opened the pull request:
# Changes
* Create new CPU-Memory-Duration function to model the serverless function's execution time based on the CPU and Memory allocation.
* Create new classes to support 2D search for the CPU-Memory space.
* Use Google Cloud `functions_v2` API to programmatically update both the CPU and Memory configuration for the function being optimized.
* Add new flag `GCPv2` to enable 2D optimization.
* Add unit tests for the new Classes.

Commit:
make modifications to optimize for both CPU and memory configurations for GCP Cloud Functions V2

Commit:
resolve merge conflicts

Commit:
fix the path naming

arshiamoghimi reviewed (changes requested):
The code looks good. Thanks. Other than the minor comments inlined, here are some items you need to fix:
- add documentations in the READMEs (main and configuration at src/configuration/README.md).
- pull the latest version of the main branch
- Fix the merge conflicts

Commit:
pull latest mainline changes, address pull request comments

hdu3 commented on src/configuration/defaults.py:
```diff
@@ -5,8 +5,8 @@
-TERMINATION_THRESHOLD = 3
+MAX_TOTAL_SAMPLE_COUNT = 250
+MIN_SAMPLE_PER_CONFIG = 5
+TERMINATION_THRESHOLD = 4
```
Thanks for the call out. Changed.

hdu3 commented on src/exploration/explorer.py:
```diff
@@ -31,6 +31,7 @@ def __init__(
         self.payload = payload
 
         self.memory_space = np.array(list(memory_space), dtype=int)
+        # self.cpu_memory_space = np.array()
```
Removed.

hdu3 commented on src/exploration/explorer_2d.py:
```diff
@@ -0,0 +1,153 @@
+                        error = e
+                    # self.cost += self.price_calculator.calculate_price(
+                    #     self._memory_config_mb, e.duration_ms, self._cpu_config
+                    # )
```
Removed.

hdu3 commented on src/exploration/explorer_2d.py:
```diff
@@ -0,0 +1,153 @@
+        #     self.price_calculator.calculate_price(
+        #         self._memory_config_mb, np.array(results), self._cpu_config
+        #     )
+        # )
```
Removed.

hdu3 commented:
> The code looks good. Thanks. Other than the minor comments inlined, here are some items you need to fix:
> 
> * add documentations in the READMEs (main and configuration at src/configuration/README.md).
> * pull the latest version of the main branch
> * Fix the merge conflicts

Thanks for the comment. I've addressed these comments in the new commit.
